types/styled-components/index.d.ts

```diff
@@ -56,6 +56,13 @@ type ReactDefaultizedProps<C, P> = C extends { defaultProps: infer D; }
     ? Defaultize<P, D>
     : P;
 
+// Omit props from P that are also contained in O. Follows similar logic to Defaultize
+type OmitOverlapping<P, O> = P extends any
+    ? string extends keyof P
```

Comment:
This seems like C&P from `Defaultize`? I think the original intent was to distribute this type over unions. This is already the case by using the conditional above (`P extends any`) and can therefore be omitted.

Reply:
It was C&P from `Defaultize` because without it, I get complaints from its usage down below with the `A` type parameter.

```
Type 'A' is not assignable to type 'Exclude<symbol, keyof O>'.
    Type 'string | number | symbol' is not assignable to type 'Exclude<symbol, keyof O>'.
        Type 'string' is not assignable to type 'Exclude<symbol, keyof O>'.
```

in the context

```
export type StyledComponentProps<
    // The Component from whose props are derived
    C extends keyof JSX.IntrinsicElements | React.ComponentType<any>,
    // The Theme from the current context
    T extends object,
    // The other props added by the template
    O extends object,
    // The props that are made optional by .attrs
    A extends keyof any
> = WithOptionalTheme<
    Omit<
        ReactDefaultizedProps<
            C,
            OmitOverlapping<React.ComponentPropsWithRef<C>, O> & O
        >,
        A
    > & Partial<Pick<React.ComponentPropsWithRef<C> & O, A>>,
    T
> & WithChildrenIfReactComponentClass<C>;
```

Reply:
I wasn't sure how else to work around it, but the reasoning around `Defaultize` seems like it would apply here as well.

Reply:
Yep got it. I guess if we omit `P extends any` we don't distribute over union props anymore and it squashes properties. 

Just to be clear: If you remove the `P extends any` it also fails the introduced tests?

Reply:
Yeah, it fails similarly.

```
Type 'A' is not assignable to type 'Exclude<keyof StyledComponentPropsWithRef<C>, keyof O>'.
  Type 'string | number | symbol' is not assignable to type 'Exclude<keyof StyledComponentPropsWithRef<C>, keyof O>'.
    Type 'string' is not assignable to type 'Exclude<keyof StyledComponentPropsWithRef<C>, keyof O>'.
```